### ISSUE TITLE
Observability and metrics - Surface metrics prometheus friendly

### ISSUE THREAD

**Issue by chrismeyersfsu** [ansible/awx#9019]
Websockets contain an [example](https://github.com/ansible/awx/blob/devel/awx/main/analytics/broadcast_websocket.py#L121) of Prometheus data structures serialized, deserialized & [exposed via a Prometheus endpoint](https://github.com/chrismeyersfsu/awx/commit/b58c71bb74bd5f5c836f88daa3d7806ff68cea47#diff-497de0f3f551df8d0333aa7a3d2cae353a08d4e2b500ca122195bafda82f6d3b).  We want to do the same for our generalized metrics. Note that websockets didn't have the potential performance needs that this metrics tracking system will need. I'm unsure of the performance of storing the metrics in Prometheus objects and _then_ serializing them. Also, the websockets stats objects persist in memory for a long time. There are some metrics workloads where the metric object will be short lived in memory. So there will be many deserialize, modify, serialize operations. For this reason, we may have our own minimal data structure that we operate on (i.e. serialize, increment, deserialize) and only convert it to a Prometheus style data-structure when rendering in the API.

"Normal" prometheus usage is to have a `/metrics` endpoint per-node that you are monitoring. We are deviating from the norm and centralizing information about multiple nodes. It would be nice to have a filter so that you get the best of both worlds, centralized metrics and the ability to use existing tooling. To this end, a filter that supports filtering by-node would be useful i.e. `/metrics?node=node1.awx.mydomain.com`

We also have to consider performance. This new set of metrics should render very quickly. Our existing metrics can be slow since they require database queries. We should use filtering or a different endpoint to allow users to avoid the pre-existing DB query metrics.

**Comment by kdelee** [comment]
- add testing for filtering in tower-qa
- add new attributes to existing tests for existence/correctness
  - basic correctness,example: num events monotonically increasing
- in cluster installs, verify data is propagated across cluster
  - example, if job is run on other node, show on other node that job count increased on job running node, and that event count increased
- add baseline pass/fail API response time benchmark test to nightly tests
  - if it takes more than a second, fail no if ands or buts about it

NEED from @fosterseth @chrismeyersfsu 

- need list of each metric, and which ones are multi-node metrics are
- discuss each metric and reasonable test case for each we want a automated correctness test

**Comment by chrismeyersfsu** [comment]
Current thoughts below. Will add other subsystems to this table.

Similar to the job lifecycle feature, some of these metrics are going to be harder to capture and may require extra state code. Use your judgement, if you hit those hard cases, skip it and come back.

### Derivable Metrics

| Higher Level Metric                         | Formula                                                      |
| ------------------------------------------- | ------------------------------------------------------------ |
| Avg seconds to insert an event              | `events_inserted_db_time` / `events_inserted`                |
| Events inserted per minute                  | `events_inserted / (events_inserted_db_time / 60)` assumes time in seconds. |
| Avg. event size in bytes                    | `events_size_bytes` / `events_inserted`                      |
| Avg seconds an event spends in memory queue | `events_memory_queue_time` / `events_in_memory` assumes time in seconds. |

### Raw Metrics

In the subsystem cell I've added a usefulness or priority score: **High, Medium, Low**

| Subsystem                            | Metric                        | Description                                                  | Thoughts                                                     |
| ------------------------------------ | ----------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| callback receiver<br />**High**      | events_inserted               | **Counter**<br />Number of individual events AFTER inserted into the database. | [Metric Location](https://github.com/ansible/awx/blob/909c2513ac6a51934cb1ad130c50dd52c4c36113/awx/main/dispatch/worker/callback.py#L110) `len(events)` would be the batch size<br /><br />We probably don't need the per-job-type breakdown that we provide in `awx-manage callback_stats`.<br />This metric has been useful in the past to feel out if maybe our callback reciever is misbehaving or if the database is slow.<br /><br />Ideally we want to update the metrics after each insert in serial mode. HOWEVER, if it's expensive ... what is considered expensive? Relative to the speed of a single serial insert. If updating the redis metric is say a factor of 10 faster than the database insert we DO NOT want to update  the metric on every serial insert. |
| callback receiver<br />**High**      | events_inserted_db_time<br /> | **Counter**<br />Time spent saving events to DB              | [Metric Location Bulk](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L116)<br />[Metric Location Serial](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L123)<br />If this is high, database slowdown suspected. Or maybe our batch size isn't optimal or batching isn't being used at all. Check `events_batch_insert_size` |
| callback receiver<br />**High**      | events_batch_save_errors      | **Counter**<br />Number of times batch insert failures. When this happens we would expect the callback receiver performance to drop due to falling back to serial insert. | [Metric Location](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L117) |
| callback receiver<br />**High**      | events_batch_size             | **Quantile**<br />I can't think of how to break this up. We really want to know the "range" of batch sizes. This seems to need a histogram or quantile data type. | [Metric Location](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L108) |
| callback receiver(ish)<br />**High** | events_insert_redis           | **Counter**<br />Similar to events_inserted_db.              | tasks.py `event_handler()` calls [Metric Location](https://github.com/ansible/awx/blob/909c2513ac6a51934cb1ad130c50dd52c4c36113/awx/main/queue.py#L35)<br />This number increases as ansible generates events. |
| callback reciever<br />**High**      | events_popped_redis           | **Counter**<br />Popped over time will tell the rate we are pulling from redis. | [Metric Location](https://github.com/ansible/awx/blob/devel/awx/main/dispatch/worker/callback.py#L59)<br />This number increases as `events_insert_redis` increases. |
| callback reciever<br />**Low**       | events_in_memory              | **Guage**<br />Added to when events are popped off Redis, subtracted from right before insertion into the DB. | [Metric Location ++](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L59)<br />[Metric Location -- If Bulk](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L116)<br />[Metric Location -- If Serial](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L123) |
| callback receiver<br />**Low**       | events_memory_queue_time      | **Guage**<br />Total time between right after an event is read from redis and when it begins to be saved to the database.<br />`memory_queue_time` / `events_in_memory` is the average time an event is in the memory queue | If this is high, then our callback receiver processing code is suspect. Maybe our batch delay mechanism isn't flushing.<br /><br />**Note:** When implementing this we want to subtract the time that each event sits in memory when it moves out of the "memory" pipeline i.e. right before it is inserted into the DB. This implies maybe a hash table where `<key=event_id, value=timestamp_popped_from_redis>` |
| callback receiver<br />**Low**       | events_size_bytes             | **Counter**<br />Total size in bytes of all events processed | [Metric Location `len(e.event_data)`](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L126)<br /><br />Observed "poor performance" of the callback receiver could be because the events are of unexpected large size. `events_size_bytes` / `events_insert_db` gives the average event size. 2KB is a normalish average |
| callback receiver<br />**Medium**    | events_queue_size_redis       | **Gauge**<br />The event queue size in redis                 | [Metric Location](https://github.com/ansible/awx/blob/devel/awx/main/dispatch/worker/callback.py#L56) <br />`LLEN` command should give us the length. |
|                                      |                               |                                                              |                                                              |
|                                      |                               |                                                              |                                                              |
|                                      |                               |                                                              |                                                              |
|                                      |                               |                                                              |                                                              |

**Comment by kdelee** [comment]
@chrismeyersfsu I'm not understanding


callback receiver | events_insert_db_time | Counter Time spent saving events to DB events_insert_db / events_insert_db_time = average time to insert | If this is high, database slowdown suspected. Or maybe our batch size isn't optimal or batching isn't being used at all. Check events_batch_insert_size
-- | -- | -- | --


are you saying that `events_insert_db_time` will just be a counter of time spend saving events, but then we could use it for further calculation to know the average time to insert? it seems like you are describing `average_event_insertion_time` in the "thoughts" column but that would have to be a 3rd thing calculated from `events_insert_db_time` and `events_insert_db` -- because it is fine if `events_insert_db_time` is high if `events_insert_db` is also high.

**Comment by chrismeyersfsu** [comment]
Correct @kdelee But we don't want to compute averages in our raw metrics. That should be computed outside, in Prometheus or Grafana. By doing this outside the system we don't have to pick a single time window. We can compute the average for the last 5 seconds, 10 minutes, or hour.

**Comment by fosterseth** [comment]
`events_size_bytes`

trying to decide a way to measure this. maybe `len(event.stdout)`?

**Comment by chrismeyersfsu** [comment]
The [`CallbackWorker`](https://github.com/ansible/awx/blob/devel/awx/main/dispatch/worker/callback.py#L30) inherits from [`BaseWorker`](https://github.com/ansible/awx/blob/devel/awx/main/dispatch/worker/base.py#L164) `work_loop` is what it sounds like, basically the `while True:` loop. It calls `read()` then `perform_work()`. Both of these functions are overridden by `CallbackWorker`. `read()` pops off redis. `perform_work()` converts the data gotten from redis, basically a blob, into a class of the proper event type. `perform_work()` then pushes that event onto [`self.buff`](https://github.com/ansible/awx/blob/6e8c71a231482eb7b7d213006efe3a316b133d50/awx/main/dispatch/worker/callback.py#L182) and triggers a flush if the requirements are met.

The above is all to say, as you have observed, we have a proper Event class at the point we want to calculate the size of the event. This "point" is right after the event is saved to the database. The attribute we want is `len(event.event_data)`.

TL;DR `len(event.event_data)`

**Comment by ryanpetrello** [comment]
Closing, cc @kdelee @fosterseth
